Scripts/static-analyzer/Sources/Analyzer/SourceNavigator.swift

```diff
@@ -0,0 +1,160 @@
+    /// the project's derived data.
+    public init(compilerInvocations: CompilerInvocations, indexStore: IndexStoreDB) {
+        self.compilerInvocations = compilerInvocations
+        self.indexStore = indexStore
+    }
+
+    public func callSites(of symbol: Symbol) -> [SymbolLocation] {
+        callSites(of: USR(rawValue: symbol.usr)!)
+    }
+
+    public func callSites(of usr: USR) -> [SymbolLocation] {
+        indexStore.occurrences(ofUSR: usr.rawValue, roles: .call)
+            .map(\.location)
+    }
+
+
+    ///Return all type-checked expressions at the given source code location.
+    ///
+    /// Take a simple map function call expression as an example.
+    ///
+    /// ```
+    /// [1, 2, 3].map { $0.description }
+    ///           -1-
+    /// ------2------
+    /// ---------------3----------------
+    /// ```
+    ///
+    /// As indicated in the code above, given a `SymbolLocation` at the `map` function call, this function returns
+    /// three expressions. The expressions are (the dash lines in the example are each expression's "byte range"):
+    /// 1. The `map` function: `([Int]) -> ((Int) throws -> String) throws -> [String]`.
+    /// 2. The `Array.map` function: `((Int) throws -> String) throws -> [String]`.
+    /// 3. The return type of the whole `map` function call: `[String]`.
+    ///
+    /// Note that all of the above are specialized signature, rather than the declared generic type.
+    public func expressions(at location: SymbolLocation) throws -> [Expression] {
+        let compilerArguments = compilerInvocations.compilerArguments(forFileAt: location.path)
+        let targetByteOffset = try Int64(StringView(String(contentsOfFile: location.path)).byteOffset(forLine: Int64(location.line), bytePosition: Int64(location.utf8Column))!.value)
+
+        let request = Request.expressionType(file: location.path, compilerArguments: compilerArguments)
+        let response = try request.send()
+
+        guard let expressionTypeList = response["key.expression_type_list"] as? [[String: SourceKitRepresentable]] else {
+            throw SourceKitResponseError.unexpectedType(key: "key.expression_type_list")
+        }
+
+        return try expressionTypeList
+            .map { try Expression(sourceKitExpressionTypeResponse: $0) }
+            .filter { $0.byteRange.contains(targetByteOffset) }
+            .sorted { $0.byteRange.count < $1.byteRange.count }
+    }
+
+    public func isInheritence(subclass subclassName: String, superclass superclassName: String, usedAt location: SymbolLocation? = nil) throws -> Bool {
+        do {
+            let subclass = try resolveType(named: subclassName)
+            let superclass = try resolveType(named: superclassName)
+            return superclasses(of: subclass).map { $0.usr }.contains(superclass.usr)
+        } catch {
+            print(error)
+        }
+
+        if let location {
+            print("Try SourceKit")
```

Comment:
I can see these `print` logs being a bit cryptic.

What do you think of something like:

```suggestion
            print("Failed to resolve type(s) for subclass \(subclassName) and superclass \(superclassName). Failed with: \(error))
        }

        if let location {
            print("Trying SourceKit to check inheritance...")
```